server/src/main/java/org/apache/druid/query/RetryQueryRunner.java

```diff
@@ -0,0 +1,204 @@
+    );
+    if (remainingResponses > 0) {
+      throw new ISE("Failed to check missing segments due to missing responds from [%d] servers", remainingResponses);
+    }
+
+    final Object maybeMissingSegments = context.get(ResponseContext.Key.MISSING_SEGMENTS);
+    if (maybeMissingSegments == null) {
+      return Collections.emptyList();
+    }
+
+    return jsonMapper.convertValue(
+        maybeMissingSegments,
+        new TypeReference<List<SegmentDescriptor>>()
+        {
+        }
+    );
+  }
+
+  /**
+   * A lazy iterator populating {@link Sequence} by retrying the query. The first returned sequence is always the base
+   * sequence given in the constructor. Subsequent sequences are created dynamically whenever it retries the query. All
+   * the sequences populated by this iterator will be merged (not combined) with the base sequence.
+   *
+   * The design of this iterator depends on how {@link MergeSequence} works; the MergeSequence pops an item from
+   * each underlying sequence and pushes them to a {@link java.util.PriorityQueue}. Whenever it pops from the queue,
+   * it pushes a new item from the sequence where the returned item was originally from. Since the first returned
+   * sequence from this iterator is always the base sequence, the MergeSequence will call {@link Sequence#toYielder}
+   * on the base sequence first which in turn initializing query distribution tree. Once this tree is built, the query
+   * nodes (historicals and realtime tasks) will lock all segments to read and report missing segments to the broker.
+   * If there are missing segments reported, this iterator will rewrite the query with those reported segments and
+   * reissue the rewritten query.
+   *
+   * @see org.apache.druid.client.CachingClusteredClient
+   * @see org.apache.druid.client.DirectDruidClient
+   */
+  private class RetryingSequenceIterator implements Iterator<Sequence<T>>
+  {
+    private final QueryPlus<T> queryPlus;
+    private final ResponseContext context;
+    private Sequence<T> sequence;
+    private int retryCount = 0;
+
+    private RetryingSequenceIterator(QueryPlus<T> queryPlus, ResponseContext context, Sequence<T> baseSequence)
+    {
+      this.queryPlus = queryPlus;
+      this.context = context;
+      this.sequence = baseSequence;
+    }
+
+    @Override
+    public boolean hasNext()
+    {
+      if (sequence != null) {
+        return true;
+      } else {
+        final List<SegmentDescriptor> missingSegments = getMissingSegments(context);
+        if (missingSegments.isEmpty()) {
+          return false;
+        } else if (retryCount >= config.getNumTries()) {
+          if (!config.isReturnPartialResults()) {
+            throw new SegmentMissingException("No results found for segments[%s]", missingSegments);
+          } else {
+            return false;
+          }
+        } else {
+          LOG.info("[%,d] missing segments found. Retry attempt [%,d]", missingSegments.size(), retryCount++);
```

Comment:
Should be `++retryCount`, right? (You would want the first retry to say "Retry attempt [1]".)

However, prefix and postfix incrementing in log messages can make things harder to read, so IMO it'd be better to split this up into two separate lines.

Reply:
Oops, just copied it from the original code. Split incrementing count from the logging.